### DISCUSSION TITLE
Sharing types between server and client

### DISCUSSION THREAD

**Discussion by elie222** [vercel/next.js#51708] [App Router]
In our project we share types between the server and client. The pattern we use most often is something like this:

```ts
const projectUserQuery = z.object({ projectId: z.string() });
export type ProjectUserQuery = z.infer<typeof projectUserQuery>;
export type ProjectUserResponse = Awaited<ReturnType<typeof getProjectUser>>;

export default async function handler() {}
```

And now the `ProjectUserQuery` and `ProjectUserResponse` can be used client-side for type-safety.

1. It doesn't seem possible to use the `server-only` package with this pattern.
2. Are there any other issues we should be aware of when using this pattern that server side code is being used client side? Or does the Next build system selectively only use the TS types and not import the actual code when only types have been imported?

**Comment by nickmarca** [comment]
@elie222 That's an intriguing question. I share the same curiosity. This pattern appears to be powerful while maintaining simplicity, bypassing the need for additional libraries or even GraphQL for robust type enforcement between client and server. It would be very valuable to receive a definitive stance from the Next.js team regarding this strategy. Is it recommended? Are there any associated risks?

From my preliminary investigations, it seems that Next.js (or Webpack) only bundles code that is directly referenced within the codebase. While I may stand corrected, even importing code from files intended for server-side use won't be included in the client bundle if it's not referenced.

I conducted a test by examining the page bundle, and there's a noticeable disparity in bundle size when I reference a function intended solely for server-side execution

Page Bundle size: 346kb - referencing server-side imported function (`getPageData`):
![Screenshot 2024-03-30 at 03 56 39](https://github.com/vercel/next.js/assets/3733116/6fef6f54-005a-4f99-8019-8bd391826047)

Page Bundle size: 121kb - without referencing server-side imported function (`getPageData`):
![Screenshot 2024-03-30 at 03 55 39](https://github.com/vercel/next.js/assets/3733116/4e39479c-8573-4472-9786-15f916f4bec6)


Additionally, for added precaution, we could implement a check within the server file to ensure it isn't mistakenly evaluated on the client side:
![Screenshot 2024-03-30 at 03 59 23](https://github.com/vercel/next.js/assets/3733116/0743e646-8608-42aa-9772-ba3964e3e403)

If we do, we would get this error on the browser console:
![Screenshot 2024-03-30 at 03 58 41](https://github.com/vercel/next.js/assets/3733116/a2a1ea54-682a-4946-9000-1f815c08102b)
